halotools/mock_observables/counts_in_cells/engines/counts_in_cylinders_engine.pyx

```diff
@@ -219,9 +239,24 @@ def counts_in_cylinders_engine(double_mesh, x1in, y1in, z1in, x2in, y2in, z2in,
                                     if (dxy_sq < rp_max_squaredtmp) & (dz_sq < pi_max_squaredtmp):
                                         counts[ifirst1+i] += 1
 
-    # At this point, we have calculated our counts on the input arrays *after* sorting
-    # Since the order of counts matters in this calculation, we need to undo the sorting
-    sorted_counts = np.array(counts)
-    idx_unsorted = unsorting_indices(double_mesh.mesh1.idx_sorted)
-    return sorted_counts[idx_unsorted]
-
+                                        if c_return_indexes:
+                                            indexes[current_indexes_cnt, 0] = ifirst1+i
+                                            indexes[current_indexes_cnt, 1] = ifirst2+j
+                                            current_indexes_cnt += 1
+                                            if current_indexes_cnt == current_indexes_len:
+                                                current_indexes_len *= 2
+                                                indexes = np.resize(indexes, (current_indexes_len, 2))
```

Comment:
I'm not sure if there is a better way to do the resize.

I tried a `Pymem_Realloc` on the pointer to the start `&indexes[0]`of the memoryview and this worked fine, but then something segfaulted later (https://github.com/astropy/halotools/blob/523e7c569ae5f4c65b6aedf2c7148e6593eab4e3/halotools/utils/array_utils.py#L213 the `.astype(int)` piece). I did some reading and as I understand it now you can't (shouldn't?) realloc memoryviews like that?

My though on this is that it will be rare so the python overhead won't be too important? The other options are 1) I have misunderstood something and we can resize memoryviews 2) Use a straight C array.

Happy to do either of them but before digging too deep wanted to get your opinion

Reply:
For practical purposes, since you're not finding the calls to np.resize to be so frequent that it matters, it's fine to just live with np.resize. So comments below do not impact PR acceptance.

Re: segfault. This is perplexing; I can't quite tell what caused this issue since I didn't see what was passed to unsorting_indices with your alternate attempt. Generally, some common causes for this kind of error with python/cython interactions are 
1. mixing up long vs. int ("i8" vs. "i4"), 
2. trying to use a memoryview like a numpy array, e.g., when using the array for fancy indexing. 
Are you sure something simple like that wasn't the problem? 

The reason I ask is that it would be good to know if http://cython.readthedocs.io/en/latest/src/tutorial/memory_allocation.html is giving out-of-date advice on how to dynamically resize memoryviews. If it's not too much trouble to look up, could you point me to where you found this idiom to be deprecated or otherwise not recommended?

Reply:
I think that those docs show how to resize memory blocks manually allocated in C, rather than the memoryview class that I am using here. I think in the example the realloc is OK because that memory is explicitly not managed by python. This is also why it needs to be wrapped it in the `SomeMemory` extension class with a `__dealloc__` method so that it can be freed by python if we return it from the C layer.

The memoryview on the other hand which I allocate here https://github.com/astropy/halotools/pull/916/files#diff-67ca324f5bf87cbe1aa05f366958f7dcR116 is still managed memory and so we can't modify it from the C layer. That messes things up in the Python as things have changed underneath it without it knowing. As you can probably tell I'm not fully sure of the details here but I think it makes sense. If I allocated that as a raw C pointer (e.g. `<double*> PyMem_Malloc(number * sizeof(double))`) then we could realloc as in the example.

The downside of doing that is you need to have the `SomeMemory` extension class that frees memory, and working with C arrays is maybe a bit more tricky/less safe. This didn't seem worth it for something that wasn't hit that much. Also when I originally wrote this I understood this a lot less. Having played with it a bit in the last week I now feel like I understand it a bit better and might be tempted to do it the C way, but I think there are tradeoffs either way.